### DISCUSSION TITLE
Predicting diverse relative permeability curves for multiple networks (Please help)

### DISCUSSION THREAD

**Discussion by sinaainesaz** [PMEAL/OpenPNM#2842] [Q&A]
Hello Openpnm community,

I am working on dynamic rock typing based on relative permeability curves. I was trying to generate different networks with different pore size distribution and wettability, but I am stuck and do not know how to solve my problem. I have a function to generate the networks I am using the codes both from the relative permeability and calibrating pore sizes tutorials. I am using weibull distribution for the pore sizes and for the wettability I assign a random number between 0 and 180 for the contact angle of the oil phase and based on that I have a if statement which checks the contact angle and if it's higher than 90 it decides to put nwp as oil and if it's lower than 90 the nwp would be water in the invasion percolation. my problems are:
first: I was aiming to have porosity in the range of 0.07 to around 0.30 but I am getting all the porosities around 0.12 and I don't know which parameters I should change and play with to have diverse porosities.
second: when I am calculating the relative permeability curves both for wetting phase and non-wetting phase and plotting against the saturation of the non-wetting phase and the curves are becoming into 2 clusters(since some times I have oil as wetting phase and sometimes water as non-wetting phase) and the curves are becoming very similar to each other you can see from the plot. I would appreciate any help and recommendation to show me what should I do to have diverse plots and different porosities for the networks.

here's my pipeline:
def generate_networks(Network_N):
    
    rock_types = ['tight_rock', 'sandstone', 'carbonate']

    data = []
    relperm_data = []

    for i in range(Network_N):
        rock_type = np.random.choice(rock_types)
        print('rock_type is:', rock_type)
        
        if rock_type == 'tight_rock':
            shape = [np.random.randint(10,20) for _ in range(3)]
            spacing = np.random.uniform(0.5e-6, 20e-6)
            scale = np.random.uniform(0.05e-6, 5e-6)
            c = np.random.uniform(0.2, 2.5)
            order_of_magnitude = int(np.floor(np.log10(scale)))
            loc = 10 ** (order_of_magnitude - 1)
        
        elif rock_type == 'sandstone':
            shape = [np.random.randint(10,25) for _ in range(3)]
            spacing = np.random.uniform(5e-6, 60e-6)
            scale = np.random.uniform(3e-6, 45e-6)
            c = np.random.uniform(0.5, 3.5)
            order_of_magnitude = int(np.floor(np.log10(scale)))
            loc = 10 ** (order_of_magnitude - 1)
        
        elif rock_type == 'carbonate':
            shape = [np.random.randint(10,25) for _ in range(3)]
            spacing = np.random.uniform(3e-6, 50e-6)
            scale = np.random.uniform(0.5e-6, 55e-6)
            c = np.random.uniform(0.3, 4.5)
            order_of_magnitude = int(np.floor(np.log10(scale)))
            loc = 10 ** (order_of_magnitude - 1)
            c = np.random.uniform(0.5, 4)

        shape = [np.random.randint(15, 25) for _ in range(3)]
        #spacing = np.random.uniform(10e-9, 50e-6)

        pn = op.network.Cubic(shape=shape, spacing=spacing)

        #for scale in scales:
         #   for c in shapes:
          #      dst = spst.weibull_min(c=c, loc=1e-7, scale=scale)

        np.random.seed(0)  # Set the state of the random number generator to "0"
        pn['pore.seed'] = np.random.rand(pn.Np)
        distribution_choice = np.random.choice(['weibull', 'norm', 'lognorm'])
        if distribution_choice == 'weibull':
            dst = spst.weibull_min(c=c, loc=loc, scale=scale)
        elif distribution_choice == 'norm':
            dst = spst.norm(loc=scale, scale=c*1e-6)
        else: # 'lognorm'
            dst = spst.lognorm(s=c, scale=scale)
        pn['pore.diameter'] = dst.rvs(pn.Np)
        #scale=np.random.uniform(1e-9, 50e-6)
        print(scale)
        #c = np.random.uniform(1, 3)
        print(c)
        #dst = spst.weibull_min(c=c, loc=loc, scale=scale)

        #pn['pore.diameter'] = dst.rvs(pn.Np)

        f = op.models.geometry.pore_seed.random
        pn.add_model(propname='pore.seed',
                    model=f,
                    num_range=[0.0, 0.98])
        pn.add_model(propname='pore.diameter',
                    model=op.models.geometry.pore_size.generic_distribution,
                    func=dst,
                    seeds='pore.seed')
        pn.regenerate_models()

        pn.add_model(propname='throat.diameter_1',
                    model=op.models.misc.from_neighbor_pores,
                    prop='pore.diameter',
                    mode='min')
        pn.add_model(propname='throat.seed',
                    model=op.models.misc.from_neighbor_pores,
                    prop='pore.seed',
                    mode='min')
        pn.add_model(propname='throat.diameter_2',
                    model=op.models.geometry.throat_size.generic_distribution,
                    func=dst)

        pn.add_model(propname='throat.diameter',
                    model=op.models.misc.scaled,
                    prop='throat.diameter_1',  # This could also be 'throat.diameter_2'
                    factor=0.6,  # This could be 1.0 if no scaling is desired
                    )

        pn.add_model(propname='pore.volume', model=op.models.geometry.pore_volume.sphere)

        pn.add_model(propname='throat.length',
                    model=op.models.geometry.throat_length.spheres_and_cylinders)
        pn.add_model(propname='throat.total_volume',
                    model=op.models.geometry.throat_volume.cylinder)
        pn.add_model(propname='throat.lens_volume',
                    model=op.models.geometry.throat_volume.lens)
        pn.add_model(propname='throat.volume',
                    model=op.models.misc.difference,
                    props=['throat.total_volume', 'throat.lens_volume'])

        geo = op.models.collections.geometry.spheres_and_cylinders
        pn.add_model_collection(geo)
        pn.regenerate_models()
        #For Oil Phase
        oil = op.phase.Air(network=pn,name='oil')
        oil['pore.surface_tension'] = np.random.uniform(0.02, 0.05)
        oil['pore.contact_angle'] = np.random.randint(0, 180)
        print('contact angle is:', oil['pore.contact_angle'][0])
        angle = oil['pore.contact_angle'][0]
        print('pore.surface_tension is:', oil['pore.surface_tension'][0])
        oil['pore.density'] = np.random.uniform(650, 900)
        print('oil pore density', oil['pore.density'])
        oil['pore.viscosity'] = np.random.uniform(0.01, 0.05)
        print('oil pore viscosity', oil['pore.viscosity'])
        oil.add_model_collection(op.models.collections.phase.air)
        oil.add_model_collection(op.models.collections.physics.basic)
        oil.regenerate_models()

        # For Water Phase
        water = op.phase.Water(network=pn, name='water')
        #water['pore.density'] = np.random.uniform(995, 1000)
        print('water pore density', water['pore.density'])
        #water['pore.viscosity'] = np.random.uniform(0.00089, 0.0013)
        print('water pore viscosity', water['pore.viscosity'])
        #water['pore.surface_tension'] = np.random.uniform(0.071, 0.075)
        print('water surface tension', water['pore.surface_tension'])
        #water['pore.vapor_pressure'] = np.random.uniform(2300, 2400)
        print('water vapor pressure', water['pore.vapor_pressure'])
        water['pore.contact_angle'] = 180 - angle
        print('water contact angle', water['pore.contact_angle'])
        water.add_model_collection(op.models.collections.phase.water)
        water.add_model_collection(op.models.collections.physics.basic)
        print('after collections thing water contact angle', water['pore.contact_angle'])
        water.regenerate_models()

        # Calculate statistical features
        pore_diameters = pn['pore.diameter']
        throat_diameters = pn['throat.diameter']
        mean_pore_size = np.mean(pore_diameters)
        std_pore_size = np.std(pore_diameters)
        mean_throat_size = np.mean(throat_diameters)
        std_throat_size = np.std(throat_diameters)

        # Calculate porosity
        inlet_porosity = pn.pores('left')
        outlet_porosity = pn.pores('right')
        A = op.topotools.get_domain_area(pn, inlets=inlet_porosity, outlets=outlet_porosity)
        L = op.topotools.get_domain_length(pn, inlets=inlet_porosity, outlets=outlet_porosity)
        Vol_bulk = A * L
        Vol_void = np.sum(pn['pore.volume']) + np.sum(pn['throat.volume'])
        porosity = Vol_void / Vol_bulk
        # Defining wettability states based on the contact angle
        if 0 <= angle < 36:  # Strong water-wet
            wp, nwp = water, oil
        elif 36 <= angle < 72:  # Weak water-wet
            wp, nwp = water, oil 
        elif 72 <= angle < 108:  # Neutral wetting
            wp, nwp = np.random.choice([water, oil], size=2, replace=False)
        elif 108 <= angle < 144:  # Weak oil-wet
            wp, nwp = oil, water  
        else:  # Strong oil-wet
            wp, nwp = oil, water

        # Relative permeability
        ip = op.algorithms.InvasionPercolation(network=pn, phase=nwp)
        Finlets_init = pn.pores('left')
        Finlets=([Finlets_init[x] for x in range(0, len(Finlets_init), 2)])
        ip.set_inlet_BC(pores=Finlets)
        ip.run()

        model_mp_cond = op.models.physics.multiphase.conduit_conductance
        water.add_model(model=model_mp_cond, propname='throat.conduit_hydraulic_conductance',
                      throat_conductance='throat.hydraulic_conductance', mode='medium', regen_mode='deferred')
        oil.add_model(model=model_mp_cond, propname='throat.conduit_hydraulic_conductance',
                      throat_conductance='throat.hydraulic_conductance', mode='medium', regen_mode='deferred')

        Snwp_num=30
        flow_in = pn.pores('left')
        flow_out = pn.pores('right')
        max_seq = np.max([np.max(ip['pore.invasion_sequence']),
                          np.max(ip['throat.invasion_sequence'])])
        start = max_seq//Snwp_num
        stop = max_seq
        step = max_seq//Snwp_num
        Snwparr = []
        relperm_nwp = []
        relperm_wp = []

        for i in range(start, stop, step):
            water.regenerate_models()
            oil.regenerate_models()
            sat = sat_occ_update(network=pn, nwp=nwp, wp=wp, ip=ip, i=i)
            Snwparr.append(sat)
            Rate_abs_nwp = Rate_calc(pn, nwp, flow_in, flow_out, conductance = 'throat.hydraulic_conductance')
            Rate_abs_wp = Rate_calc(pn, wp, flow_in, flow_out, conductance = 'throat.hydraulic_conductance')
            Rate_enwp = Rate_calc(pn, nwp, flow_in, flow_out, conductance = 'throat.conduit_hydraulic_conductance')
            Rate_ewp = Rate_calc(pn, wp, flow_in, flow_out, conductance = 'throat.conduit_hydraulic_conductance')
            relperm_nwp.append(Rate_enwp/Rate_abs_nwp)
            relperm_wp.append(Rate_ewp/Rate_abs_wp)
<img width="1012" alt="Screenshot 2023-10-05 at 21 52 35" src="https://github.com/PMEAL/OpenPNM/assets/81174257/cce6402a-d41c-4f8b-ab63-26a0c28aa265">

**Comment by jgostick** [comment]
That's too much code for me to digest in the time I have, but let me address your bigger picture questions:
1. The easiest way to vary porosity is by increase/decreasing the lattice spacing.  This effectively creates more/less solid between the pores. A more difficult way is to randomly delete some pores, which would also add some solid.  You'd have to do this sparingly to not change the connectivity distribution much.  When calculating porosity you have be careful...I hope you have consulted [this example](https://openpnm.org/examples/applications/porosity.html). (Note: this example seems to have broken rendering...I will try to touch it up soon).
2. Rel perm is *normalized* by absolute perm, so any changes you make to the network that results in large changes in K, will be mostly cancelled out when you divide the 2phase K by K_abs.  So don't expect much change.  You might want to plot the K of the networks to prove to yourself that this is changing.  Changing wettability of the network will also have 0 effect on the Krel curves (assuming you are just changing contact angle everywhere).  All this does is shift the curves left and right on the Pc axis, but you are plotting against satn (the y axis).  Another way to look at this is to plot the Pc curve for each material, then divide the Pc values by the "break through pressure" (can be approximated as the Pc at s=0.5 or something like that).  All your curves will probably collapse, which is why your seeing such similar Krel. There is a reason why many people just assume that Krel = (1-s)^3....because all Krel behave the same. 

If you really want to show Krel changing you could:
1. Use a distribution of contact angles, probably with some spatial correlation.
2. Change the connectivity of the network be deleting pores and/or throats.  

Also note that openpnm does not natively support 'film flow' so the wetting phase "may" be flowing in corners in your experiment but this is not included in your simulations.  This would not be too difficult to implement yourself...you just have write a [custom pore scale model](https://openpnm.org/examples/reference/architecture/pore_scale_models.html) that calculates the hydraulic conductance of the wetting phase in throat mostly filled with non-wetting phase.

**Reply by jgostick** [reply]
I also want to add that you could generate the Pc curves from your networks, then fit the Brook-Corry and van Genuchten models to them to get the Krel curves...you see again that these fits all predict similar curves.

**Reply by sinaainesaz** [reply]
Dear Professor, 
I really appreciate your time and your help alot.
I have been working on this matter to receive much diverse relperm curves. I applied the contact angle for the oil phase using spatial correlation now the porese close to each other have similar contact angles. and also for the contact angle of the water phase I am assigning the (180-oil['pore.contact_angle']). the plot shows much more difference from the one I was receiving previously.
I wanted to know what else parameters I can change of assing random values using spatial correlation that have impact on the relperm and can predict my curves much more similar to the real world.
since I have this parameters in the oil phase.

oil : <openpnm.phase.Air at 0x2bac2ce00>
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
  #  Properties                                                   Valid Values
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
  2  pore.temperature                                              3375 / 3375
  3  pore.pressure                                                 3375 / 3375
  4  pore.density                                                  3375 / 3375
  5  pore.molar_density                                            3375 / 3375
  6  pore.diffusivity                                              3375 / 3375
  7  pore.thermal_conductivity                                     3375 / 3375
  8  pore.viscosity                                                3375 / 3375
  9  pore.surface_tension                                          3375 / 3375
 10  pore.contact_angle                                            3375 / 3375
 11  throat.entry_pressure                                         9450 / 9450
 12  throat.hydraulic_conductance                                  9450 / 9450
 13  throat.diffusive_conductance                                  9450 / 9450
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
  #  Labels                                                 Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
  2  pore.all                                                             3375
  3  throat.all                                                           9450
――――――――――――――――――――――――――

Please also if you could check out my new plots and let me know if now they make sense and are better for my rock typing analysis.

Regards,
Sina
![second](https://github.com/PMEAL/OpenPNM/assets/81174257/262877aa-3280-4e30-afdf-60d651d44991)

**Reply by jgostick** [reply]
Basically I think you should back-up one step.  Adjust things so that you get a variety of different capillary pressure curves...like different slopes, steps, etc.  If you can get these, THEN your rel perm curves will change.  You can do this by changing the pore size distribution, as described [here](https://openpnm.org/examples/applications/adjusting_pore_size_distributions.html).  Or by changing the wetttability as you've already done.  You can also get fancy with the hydraulic conductance models so that when a phase is invaded with non-wetting it still has some conductivity for the wetting phase, via film and corner flows. This is the standard trick that rock/oil people use, but we do not have this implemented in openpnm.  If this is of interest you should find some paper that derive hydraulic conductance models, then it's relatively easy to implement them in openpnm, and perhaps [this example](https://openpnm.org/examples/reference/architecture/pore_scale_models.html) will help.

**Reply by sinaainesaz** [reply]
Dear Professor gostick,

I really appreciate your helps and your answers to my questions. It's been really helpful during my simulations and building networks. we have decided to put our focus on the topology of the networks instead of applying spatial correlation for contact angles so now I am only assigning one random number between 90 and 180 for oil phase. I have been using weibull distribution for pore sizes and using the neighboring method I am adjusting also the throat sizes like the example ' Calibrating a Network Using Mercury Intrusion Data'. 
I am also applying some changes to the connectivity of the network like the example in 'Generating Networks and Quick Visualization'. That's where I found that I could change the connectivity of the pores in cubic network randomly I am assigning the connectivity parameter in the cubic one of these numbers (6, 18,26) and randomly select some throats and pores to eliminate from the networks so in this way I can make difference between my networks. this code snippet is how I am doing it:

num_throats_to_remove = np.random.randint(100, pn.Nt)
drop = np.random.choice(np.arange(pn.Nt), num_throats_to_remove, replace=False)
op.topotools.trim(network=pn, throats=drop)
num_pores_to_remove = np.random.randint(1, int(0.2*pn.Np))  # Removing up to 20% of pores
pores_to_remove = np.random.choice(np.arange(pn.Np), num_pores_to_remove, replace=False)
op.topotools.trim(network=pn, pores=pores_to_remove)

inlet_porosity = pn.pores('left')
print('len inlet left', len(inlet_porosity))
outlet_porosity = pn.pores('right')
print('len outlet right', len(outlet_porosity))
    

G = convert_to_networkx(pn)


# Check for a path from any inlet pore to any outlet pore
path_exists = any([nx.has_path(G, source=inlet, target=outlet) for inlet in inlet_porosity for outlet in outlet_porosity])

if not path_exists:
    print("No percolation from inlet to outlet. Regenerating network for this iteration.")
    continue

# Add the cluster models
pn.add_model(propname='pore.cluster_number', model=op.models.network.cluster_number)
pn.add_model(propname='pore.cluster_size', model=op.models.network.cluster_size)

# Identify the size of the main cluster
main_cluster_size = np.max(pn['pore.cluster_size'])

# Identify and trim the pores that belong to smaller clusters
Ps = pn['pore.cluster_size'] < main_cluster_size
op.topotools.trim(network=pn, pores=Ps)

If I wanna show you an example for a network I am receiving some networks like this:
![image](https://github.com/PMEAL/OpenPNM/assets/81174257/17adf2b9-2fda-42fa-b55f-0db9fd4f1ac0)

As you told me to make changes to receive diverse capillary pressure curves. by these modifications I am receiving these capillary pressure curves using drainage algorithm.
![image](https://github.com/PMEAL/OpenPNM/assets/81174257/3f0fee92-8f54-48bf-8e99-1ec0235f7089)

And here are my relperm curves for 11 different networks:
![image](https://github.com/PMEAL/OpenPNM/assets/81174257/a3b9dadb-462e-4662-b0a7-14ec641350c3)



the problem is after this random removing of pores and throats I receive two errors:
ERROR    Inlet and outlet faces are different area                                    [_topotools.py]
ERROR    A unique value of length could not be found                                  [_topotools.py]

I would appreciate a lot if you please tell me how can I solve this problem? and if this effecting the relperm curves?

Regards,
Sina

**Reply by jgostick** [reply]
The functions for finding domain area and length are approximate and only really work if the network is truly cubic.  We do a few checks in there to makes sure the necessary conditions are met.  If they are not met by your network then the solutions is to not use those functions.  The code you put above does not have enough information for me to know exactly what is causing the problem, but I am guessing you are calling the `find_domain_area` and `find_domain_length` functions.  Instead of calling these, you can just enter the values by hand since you know the initial size of the network before you did all the trimming.

**Reply by sinaainesaz** [reply]
Hi,

Yes. As I checked my pipeline in the part related to calculating the porosity I am using those two functions.
like this:
inlet_porosity = pn.pores('left')
        outlet_porosity = pn.pores('right')
        A = op.topotools.get_domain_area(pn, inlets=inlet_porosity, outlets=outlet_porosity)
        L = op.topotools.get_domain_length(pn, inlets=inlet_porosity, outlets=outlet_porosity)
        Vol_bulk = A * L
        Vol_void = np.sum(pn['pore.volume']) + np.sum(pn['throat.volume'])
        porosity = Vol_void / Vol_bulk


Since I am making changes in the connectivity and the throats and pores in the network randomly. So my networks would not be fully cubic.

Do you think if I still use these functions to calculate the porosity is fine?

Since I tried to add the A and L manually like this:

# The way I am calculating the porosity without get_domain_area and get_domain_lenght
        A = (N * spacing) ** 2
        L = N * spacing
        
        # Calculate porosity
        inlet_porosity = pn.pores('left')
        outlet_porosity = pn.pores('right')

        Vol_bulk = A * L
        Vol_void = np.sum(pn['pore.volume']) + np.sum(pn['throat.volume'])
        print('vol void after elimination', Vol_void)
        porosity = Vol_void / Vol_bulk
        
        print('My calculated porosity', porosity)



        # Calculate porosity
        inlet_porosity = pn.pores('left')
        outlet_porosity = pn.pores('right')
        A = op.topotools.get_domain_area(pn, inlets=inlet_porosity, outlets=outlet_porosity)
        L = op.topotools.get_domain_length(pn, inlets=inlet_porosity, outlets=outlet_porosity)
        Vol_bulk = A * L
        Vol_void = np.sum(pn['pore.volume']) + np.sum(pn['throat.volume'])
        porosity = Vol_void / Vol_bulk
        
        print('openpnm calculated porosity', porosity)
        

When I print out the value for porosity calulated by me and the value calculated by openpnm functions I receive this difference.

My calculated porosity 0.04225861487660535
openpnm calculated porosity 0.05957511495762031

or for another network:
My calculated porosity 0.16176410801393126
openpnm calculated porosity 0.2218986392509345

Regards,
Sina

**Comment by likeduck** [comment]
I guess your problem arises here?
![03958B0B8E144F3F147BBEB1F2A39A60](https://github.com/PMEAL/OpenPNM/assets/117707506/53ba9291-c7bc-434c-8e89-5f20789b51d1)
Although I don't understand how this func(`get_domain_area`) works, I think pruning the network affects the calculation process.

**Reply by sinaainesaz** [reply]
Hello there,

Thank you for your reply to my question. Yes, it seems that this is the case. But I was still wondering if it's still possible to use those two functions to calculate the porosity and absolute permeability since when I call those functions in the pipeline it throws the error but does not break the code. And if it's not good to use them. should I calculate the A and L like this for a network with size 10 and the spacing like this?

A = (10 * spacing) ** 2
L = 10 * spacing

# Calculate porosity
inlet_porosity = pn.pores('left')
outlet_porosity = pn.pores('right')

Vol_bulk = A * L
Vol_void = np.sum(pn['pore.volume']) + np.sum(pn['throat.volume'])
print('vol void after elimination', Vol_void)
porosity = Vol_void / Vol_bulk

print('My calculated porosity', porosity)
